### PR TITLE
Asf dispatch to botocore

### PR DESCRIPTION
**Pull request by calvernaz**
This is just enough to review the mechanics, this isn't a working example. You can find a working example [here](asf-migration-apigateway)

### CODE REVIEW

File: localstack/aws/api/core.py [hunk 1]
```diff
@@ -221,7 +221,9 @@ def service_operation(self) -> ServiceOperation:
 ServiceRequestHandler = Callable[[RequestContext, ServiceRequest], Optional[ServiceResponse]]
 
 
-def handler(operation: str = None, context: bool = True, expand: bool = True):
+def handler(
```
Comment (calvernaz) [inline localstack/aws/api/core.py:224]: The handler will have an extra property in the marker to signal the implementation is in the provider, for the cases we want to add functionality, for example, custom implementations not in moto.

File: localstack/aws/skeleton.py [hunk 2]
```diff
@@ -22,6 +24,19 @@
...
 
 
+# Cache the client to avoid instantiation per bound method. Also, they are thread-safe with some
+# caveats
+# https://boto3.amazonaws.com/v1/documentation/api/latest/guide/clients.html
+@lru_cache
+def boto_client(service_name):
+    return boto3.client(service_name)
+
+
```
Comment (calvernaz) [inline localstack/aws/skeleton.py:34]: boto setup to bound a function to `ServiceRequestDispatcher` that is not the provider implementation (non-override use case)

File: localstack/aws/skeleton.py [hunk 3]
```diff
@@ -58,16 +74,24 @@ def create_dispatch_table(delegate: object) -> DispatchTable:
...
-                    fn.__name__, fn.operation, fn.pass_context, fn.expand_parameters
+                    fn.__name__, fn.operation, fn.pass_context, fn.expand_parameters, fn.override
                 )
             except AttributeError:
                 pass
 
     # create dispatch table from operation handlers by resolving bound functions on the delegate
     dispatch_table: DispatchTable = {}
     for handler in handlers.values():
+
```
Comment (calvernaz) [inline localstack/aws/skeleton.py:85]: here is where we dispatch between provider implementation or boto.

File: localstack/services/apigateway/apigateway_starter.py [hunk 4]
```diff
@@ -543,12 +543,22 @@ def create_rest_api(self, *args, tags={}, **kwargs):
 def start_apigateway(port=None, backend_port=None, asynchronous=None, update_listener=None):
     port = port or config.PORT_APIGATEWAY
     apply_patches()
-    result = start_moto_server(
+
```
Comment (calvernaz) [inline localstack/services/apigateway/apigateway_starter.py:546]: moto mock setup needs to happen before boto usage.

**Comment by thrau** [comment]
I'm struggling to understand how the code that uses this will look like exactly. Is there a way you can distill the working example down to a couple of code snippets to make it clearer?

**Comment by calvernaz** [comment]
> I'm struggling to understand how the code that uses this will look like exactly. Is there a way you can distill the working example down to a couple of code snippets to make it clearer?

Sure, not sure which part you didn't understand, I tried to keep it to the minimum in this PR.

But essentially, the driver will be the `override` property defining if the bounded function is implemented by the provider *or* a direct call to moto (implicit).

In this example below the call is not overriden by the provider, so i'm wiring `create_rest_api` to boto.
```
class ApigatewayProvider(ApigatewayApi):
    @handler("CreateRestApi")
    def create_rest_api(...)
```

Let's say I want to override this call inside Localstack and add more functionality or add a custom implementation that doesn't uses moto.
```
class ApigatewayProvider(ApigatewayApi):
    @handler("CreateRestApi", override=True)
    def create_rest_api(...): 
       ... my implementation ...
```

Again, not sure i'm tackling your struggle.

**Comment by thrau** [comment]
thanks for clarifying.

> In this example below the call is not overriden by the provider, so i'm wiring `create_rest_api` to boto.

is this now the default behavior, i.e., calls being dispatched to moto? that would seem to be a rather strong assumption for this level of abstraction.

**Comment by calvernaz** [comment]
> thanks for clarifying.
> 
> > In this example below the call is not overriden by the provider, so i'm wiring `create_rest_api` to boto.
> 
> is this now the default behavior, i.e., calls being dispatched to moto? that would seem to be a rather strong assumption for this level of abstraction.

In this example yes, was the default. But unless moto doesn't implement the majority of the functionality I'd not consider the opposite as the default - and then it begs the question why are we using moto.

**Pull request closed without merging**